### PR TITLE
feat(hdwallet-keepkey): add eip-712 signtypeddata support

### PR DESCRIPTION
**Pull request by pastaghost**

### CODE REVIEW

**Review by elmutt — DISMISSED** [review]

File: packages/hdwallet-keepkey/src/transport.ts [hunk 1]
```diff
@@ -92,7 +92,13 @@ export class Transport extends core.Transport {
     const buffer = new Uint8Array(9 + 2 + msgLength);
     buffer.set(first.slice(0, Math.min(first.length, buffer.length)));
 
-    for (let offset = first.length; offset < buffer.length; ) {
+    // msgLength includes the preamble overhead "?" for each chunk.
+    // We don't want the preamble from each chunk in the buffer (though, curiously, the preamble 
+    // for the first chunk is included in the buffer for some reason).
+    // We do, however, have to count the preamble in each chunk to make the proper number
+    // of chunked reads
+    let msgLenReceived = 64 // already have first buffer
```
Comment (BitHighlander) [inline packages/hdwallet-keepkey/src/transport.ts:100]: 👏

**Review by BitHighlander — DISMISSED** [review]
exciting 🚢

File: packages/hdwallet-keepkey/src/bitcoin.ts [hunk 2]
```diff
@@ -70,36 +70,7 @@ function prepareSignTx(
     const output: core.BTCSignTxOutput = o;
     const newOutput = new Types.TxOutputType();
     newOutput.setAmount(Number(output.amount));
-    if (output.exchangeType) {
```
Comment (kaladinlight) [inline packages/hdwallet-keepkey/src/bitcoin.ts:73]: seems like unrelated changes. was this related to legacy shapeshift exchange?

File: packages/hdwallet-keepkey/src/bitcoin.ts [hunk 3]
```diff
@@ -112,6 +83,7 @@ function prepareSignTx(
     } else {
       // BTCSignTxOutputSpend
       newOutput.setScriptType(Types.OutputScriptType.PAYTOADDRESS);
+      assert(output.address != undefined, "Output must have a valid BTC address.");
```
Comment (kaladinlight) [inline packages/hdwallet-keepkey/src/bitcoin.ts:86]: preference for strict equality checks. could also just do a falsey `!output.address` check as it is really checking if value exists and not necessarily if valid or not (could update error string to reflect this as well).
```suggestion
      assert(output.address !== undefined, "Output must have a valid BTC address.");
```

File: packages/hdwallet-keepkey/src/bitcoin.ts (hunk 2 above)
Comment (kaladinlight) [reply packages/hdwallet-keepkey/src/bitcoin.ts:73]: if not supported anymore, I would clean up any related types as well and just make a separate PR for this removal

File: packages/hdwallet-keepkey/src/ethereum.ts [hunk 4]
```diff
@@ -1,14 +1,16 @@
+/* eslint-disable no-debugger */
```
Comment (kaladinlight) [inline packages/hdwallet-keepkey/src/ethereum.ts:1]:
```suggestion
```

File: integration/src/ethereum/ethereum.ts [hunk 5]
```diff
@@ -421,5 +423,54 @@ export function ethereumTests(get: () => { wallet: core.HDWallet; info: core.HDW
       },
       TIMEOUT
     );
+
+    test.skip(
```
Comment (kaladinlight) [inline integration/src/ethereum/ethereum.ts:427]:
```suggestion
    test(
```

File: integration/src/ethereum/ethereum.ts [hunk 6]
```diff
@@ -1,4 +1,6 @@
+/* eslint-disable jest/no-disabled-tests */
```
Comment (kaladinlight) [inline integration/src/ethereum/ethereum.ts:1]:
```suggestion
```

File: packages/hdwallet-keepkey/src/ethereum.ts [hunk 7]
```diff
@@ -188,21 +173,116 @@ export async function ethGetAddress(transport: Transport, msg: core.ETHGetAddres
...
+  transport: Transport,
+  msg: core.ETHSignTypedData
+): Promise<core.ETHSignedTypedData> {
+  /**
+   * If the message to be signed is sufficiently small, the KeepKey can calculate the
+   * domain separator and message hashes. Otherwise, we need to pre-calculate hashes
+   * here and verify on device.
+   */
+
+  let result: Ethereum.EthereumTypedDataSignature;
```
Comment (kaladinlight) [inline packages/hdwallet-keepkey/src/ethereum.ts:199]: not being reassigned within logic forks, may as well just assign as const locally scoped

File: packages/hdwallet-keepkey/src/ethereum.ts [hunk 8]
```diff
@@ -188,21 +173,116 @@ export async function ethGetAddress(transport: Transport, msg: core.ETHGetAddres
...
+
+  const sTypes = JSON.stringify({ types: msg.types });
+  const sPrimaryType = JSON.stringify({ primaryType: msg.primaryType });
+  const sDomain = JSON.stringify({ domain: msg.domain });
+  const sMessage = JSON.stringify({ message: msg.message });
+
+  try {
+    if (sTypes.length > 2048 || sPrimaryType.length > 80 || sDomain.length > 2048 || sMessage.length > 2048) {
+      /* Pre-calculate domain separator and messages hashes and verify on KeepKey */
+      const hashableMessage: TypedData = {
```
Comment (kaladinlight) [inline packages/hdwallet-keepkey/src/ethereum.ts:209]: thoughts about nesting this object in `coreETHSignTypedData` to prevent the need to recreate the `TypedData` object?
```
export type ETHSignTypedData = { data: TypedData; addressNList: BIP32Path };
```

File: packages/hdwallet-keepkey/src/ethereum.ts [hunk 9]
```diff
@@ -188,21 +173,116 @@ export async function ethGetAddress(transport: Transport, msg: core.ETHGetAddres
...
+      let messageHash = undefined;
+      const domainSeparatorHash = getTypeHash(hashableMessage, "EIP712Domain");
+      const domainSeparatorHash64 = Buffer.from(domainSeparatorHash).toString("base64");
+
+      const t = new Ethereum.EthereumSignTypedHash();
+      t.setAddressNList(msg.addressNList);
+      t.setDomainSeparatorHash(domainSeparatorHash64);
+
+      if (msg.message) {
+        messageHash = getMessage(hashableMessage, true);
```
Comment (kaladinlight) [inline packages/hdwallet-keepkey/src/ethereum.ts:224]: message is required on `TypedData` and while testing, it looks like `getTypeHash` throws if the payload does not conform correctly as is, don't think we need this check and can get rid of the `let` above

File: packages/hdwallet-keepkey/src/ethereum.ts [hunk 10]
```diff
@@ -188,21 +173,116 @@ export async function ethGetAddress(transport: Transport, msg: core.ETHGetAddres
...
+      let response = await transport.call(Messages.MessageType.MESSAGETYPE_ETHEREUM712TYPESVALUES, dsh, {
+        msgTimeout: core.LONG_TIMEOUT,
+        omitLock: true,
+      });
+
+      const mh = new Ethereum.Ethereum712TypesValues();
+      mh.setAddressNList(msg.addressNList);
+      mh.setEip712types(sTypes);
+      mh.setEip712primetype(sPrimaryType);
+
```
Comment (kaladinlight) [inline packages/hdwallet-keepkey/src/ethereum.ts:260]:
```suggestion
```

File: packages/hdwallet-keepkey/src/ethereum.ts [hunk 11]
```diff
@@ -188,21 +173,116 @@ export async function ethGetAddress(transport: Transport, msg: core.ETHGetAddres
...
+      mh.setEip712data(sMessage);
+      mh.setEip712typevals(2);
```
Comment (kaladinlight) [inline packages/hdwallet-keepkey/src/ethereum.ts:262]:
```suggestion
      mh.setEip712typevals(2);

```

File: packages/hdwallet-keepkey/src/transport.ts [hunk 12]
```diff
@@ -92,7 +92,13 @@ export class Transport extends core.Transport {
...
     buffer.set(first.slice(0, Math.min(first.length, buffer.length)));
 
-    for (let offset = first.length; offset < buffer.length; ) {
+    // msgLength includes the preamble overhead "?" for each chunk.
+    // We don't want the preamble from each chunk in the buffer (though, curiously, the preamble
+    // for the first chunk is included in the buffer for some reason).
+    // We do, however, have to count the preamble in each chunk to make the proper number
+    // of chunked reads
+    let msgLenReceived = 64; // already have first buffer
+    for (let offset = first.length; msgLenReceived < msgLength; msgLenReceived += 64) {
```
Comment (kaladinlight) [inline packages/hdwallet-keepkey/src/transport.ts:101]: weird ordering of the for loop vars
```suggestion
    let offset = first.length;
    for (let msgLenReceived = 64; msgLenReceived < msgLength; msgLenReceived += 64) {
```

File: packages/hdwallet-keepkey/src/ethereum.ts (hunk 8 above)
Comment (pastaghost) [reply packages/hdwallet-keepkey/src/ethereum.ts:209]: sTypes, sPrimaryType, sDomain, sMessage are representations of the data in the ETHSignedTypedData object formatted specifically for KeepKey. Personally, I like to keep application-specific logic as close to the application as possible. Here, I think it's cleaner to transform the data for KeepKey in the hdwallet-keepkey adapter instead of extending the shared data type to include fields that are only consumed in one place.

Comment (kaladinlight) [reply packages/hdwallet-keepkey/src/ethereum.ts:209]: Just to clarify, I was meaning this comment to be in reference to `hashableMessage` vs the `sVariables`. 
image

Comment (pastaghost) [reply packages/hdwallet-keepkey/src/ethereum.ts:209]: We aren't yet using this type anywhere else, and there isn't any reason why the object necessarily needs to be flat, so I don't see any issue with `core.ETHSignTypedData` having separate fields for the hashable data and addressNList. I'll make that change now.

Comment (kaladinlight) [reply packages/hdwallet-keepkey/src/ethereum.ts:209]: Feels more ergonomic this way to me. Cool, thanks!

File: examples/sandbox/index.ts [hunk 13]
```diff
@@ -1819,6 +1822,36 @@ $ethEIP1559.on("click", async () => {
...
 
+$ethSignTypedData.on("click", async (e) => {
+  e.preventDefault();
+  if (!wallet) {
+    $ethResults.val("No wallet?");
+    return;
+  }
+  if (core.supportsETH(wallet)) {
+    const result = await wallet.ethSignTypedData({
+      ...eip712["calculateHashesOnDevice"],
```
Comment (kaladinlight) [inline examples/sandbox/index.ts:1833]: will have to update the sandbox payloads for `hashableData` as well so we don't forget

File: packages/hdwallet-core/src/ethereum.ts [hunk 14]
```diff
@@ -107,6 +108,18 @@ export interface ETHVerifyMessage {
   signature: string;
 }
 
+export type ETHSignTypedData = {
+  hashableData: TypedData;
+  addressNList: BIP32Path;
+};
+
+export type ETHSignedTypedData = {
```
Comment (kaladinlight) [inline packages/hdwallet-core/src/ethereum.ts:116]: @pastaghost is there a reason for returning the `domainSeparatorHash` and `messageHash` as opposed to just the address + sig like in `signMessage`?
